Asset bundles bleedingedge

chloelcdev opened the pull request.

chloelcdev committed:
Turned AssetBundleHelper into AssetLoader, added WrappedAssetBundle

also added WrappedAssetBundleRequest

chloelcdev committed:
Explicitly specify type for mono

chloelcdev committed:
remove deprecated "mainAsset" from WrappedAssetBundle

chloelcdev committed:
Add back line I guess I deleted? always use UnityEngine.Object in WrappedAssetBundle

chloelcdev committed:
take away try/catch

chloelcdev committed:
Add Dependencies for AssetBundle things

(MelonLoaderAssembliesPath) \UnityEngine.Il2CppAssetBundleManager.dll
(MonoAssembliesPath) \UnityEngine.AssetBundleModule.dll

chloelcdev committed:
Create AssetBundleHelper.cs

chloelcdev committed:
Fix quotation marks in csproj

chloelcdev committed:
Update AssetLoader.cs

chloelcdev committed:
Create UpdateBepInExBundleManager.bat

chloelcdev committed:
Have batch file wrap #if IL2CPPBEPINEX tags

chloelcdev committed:
Add downloaded BE.Il2CppAssetBundleManager

chloelcdev committed:
Merge branch 'bleeding-edge' into AssetBundles-bleedingedge

MaxtorCoder commented on S1API/AssetBundles/AssetLoader.cs:
```diff
@@ -0,0 +1,88 @@
+        ///         GameObject myCar = Instantiate(EasyLoad<GameObject>("bundles.carassetbundle", "MyCarGameObject"));
+        ///         
+        /// </summary>
+        public static T EasyLoad<T>(string bundle_name, string object_name) where T : UnityEngine.Object
+        {
+            return EasyLoad<T>(bundle_name, object_name, Assembly.GetExecutingAssembly(), out _);
+        }
+
+        public static T EasyLoad<T>(string bundle_name, string object_name, out WrappedAssetBundle bundle) where T : UnityEngine.Object
+        {
+            return EasyLoad<T>(bundle_name, object_name, Assembly.GetExecutingAssembly(), out bundle);
+        }
+
+        public static T EasyLoad<T>(string bundle_name, string object_name, Assembly assemblyOverride) where T : UnityEngine.Object
+        {
+            return EasyLoad<T>(bundle_name, object_name, assemblyOverride, out _);
+        }
+
+        public static T EasyLoad<T>(string bundle_name, string object_name, Assembly assemblyOverride, out WrappedAssetBundle bundle) where T : UnityEngine.Object
+        {
+            // Get the asset bundle from the assembly
+            bundle = GetAssetBundleFromStream($"{assemblyOverride.GetName().Name}.{bundle_name}");
+
+            // Load the asset from the bundle
+            return bundle.LoadAsset<T>(object_name);
+        }
```
Convert variable naming to `camelCase`

MaxtorCoder commented on S1API/AssetBundles/WrappedAssetBundle.cs:
```diff
@@ -0,0 +1,209 @@
+﻿using UnityEngine;
+
+#if IL2CPP
+using Type = Il2CppSystem.Type;
+using AssetBundle = UnityEngine.Il2CppAssetBundle;
+using AssetBundleRequest = UnityEngine.Il2CppAssetBundleRequest;
+#else
+using Type = System.Type;
+#endif
+
+using Object = UnityEngine.Object;
```
Why usings like this? Does it not work otherwise?

MaxtorCoder commented on S1API/AssetBundles/WrappedAssetBundle.cs:
```diff
@@ -0,0 +1,209 @@
+        {
+            _realBundle = realBundle;
+        }
+
+
```
Remove white space here and maybe convert the methods to lambda's?
E.g.
```cs
public bool Contains(string name) => _realBundle.Contains(name);
```

MaxtorCoder commented on S1API/AssetBundles/WrappedAssetBundle.cs:
```diff
@@ -0,0 +1,209 @@
+    /// <summary>
+    ///     Works just like the AssetBundleRequest type, it's just a proxy, but will use Il2CppAssetBundleRequest if it needs to so you don't have to worry about it.
+    /// </summary>
+    public class WrappedAssetBundleRequest
+    {
+        public AssetBundleRequest _realRequest;
+
+        public WrappedAssetBundleRequest(AssetBundleRequest realRequest)
+        {
+            _realRequest = realRequest;
+        }
+
+        public Object asset => _realRequest.asset;
+
+        public Object[] allAssets => _realRequest.allAssets;
+    }
```
Would be cleaner to separate this class into it's own file

MaxtorCoder commented on S1API/AssetBundles/WrappedAssetBundle.cs:
```diff
@@ -0,0 +1,209 @@
+        string[] managedStrings = new string[il2cppStrings.Length];
+        for (int i = 0; i < il2cppStrings.Length; i++)
+        {
+            managedStrings[i] = il2cppStrings[i];
+        }
+        return managedStrings;
+    }
+}
+
+public static class Il2CppObjectArrayExtensions
+{
+    public static T[] ToManagedArray<T>(this Il2CppReferenceArray<T> il2cppObjects) where T : Object
+    {
+        T[] managedObjects = new T[il2cppObjects.Length];
+        for (int i = 0; i < il2cppObjects.Length; i++)
+        {
+            managedObjects[i] = il2cppObjects[i];
+        }
+        return managedObjects;
+    }
+}
+*/
+
+
+/*
+ * 
+ * [12:57:44.463] [ManorMod] Unhandled exception in coroutine. It will not continue executing.
+System.MissingMethodException: Method not found: 'UnityEngine.AssetBundle UnityEngine.AssetBundle.LoadFromStream(System.IO.Stream)'.
+   at S1API.AssetBundles.AssetLoader.GetAssetBundleFromStream(String fullResourceName)
+   at S1API.AssetBundles.AssetLoader.EasyLoad[T](String bundle_name, String object_name, Assembly assemblyOverride, WrappedAssetBundle& bundle)
+   at S1API.AssetBundles.AssetLoader.EasyLoad[T](String bundle_name, String object_name)
+   at ManorMod.Core.LoadAssetBundle()+MoveNext()
+   at MelonLoader.Support.MonoEnumeratorWrapper.MoveNext() in D:\a\MelonLoader\MelonLoader\Dependencies\SupportModules\Il2Cpp\MonoEnumeratorWrapper.cs:line 39
+[12:57:48.792] [ManorMod] System.MissingMethodException: Method not found: 'Void UnityEngine.Events.UnityAction..ctor(System.Object, IntPtr)'.
+   at ManorMod.Core.OnLateInitializeMelon()
+   at MelonLoader.MelonEvent.<>c.<Invoke>b__1_0(LemonAction x) in D:\a\MelonLoader\MelonLoader\MelonLoader\Melons\Events\MelonEvent.cs:line 174
+   at MelonLoader.MelonEventBase`1.Invoke(Action`1 delegateInvoker) in D:\a\MelonLoader\MelonLoader\MelonLoader\Melons\Events\MelonEvent.cs:line 143
+
+
+*/
```
Add a README.md file for this in the AssetBundles file maybe?

MaxtorCoder committed:
fix: Some code style changes

TODO:
- Track all loaded asset bundles (?)
- Test BepInEx once it get's merged properly

MaxtorCoder committed:
Rename back to AssetLoader

MaxtorCoder committed:
Merge branch 'bleeding-edge' into AssetBundles-bleedingedge

MaxtorCoder committed:
fix: Fix for BepInEx building

MaxtorCoder committed:
fix: Add AssetBundle caching and fix `overrideAssembly` parameter
